### ISSUE TITLE
[IREE] Ben's list of examples he wishes to have

### ISSUE THREAD

**Issue by josemonsalve2** [ypapadop-amd/HSA-examples#10]
This came out of a discussion with Ben Vanik from IREE. 

```
Research:
  HIP <-> HSA interop (buffers/synchronization).
  HSA multi-device (buffers/synchronization).
  intra-queue concurrency (same device, same queue).
  inter-queue concurrency (same device, different queues).
  intra-queue device-side enqueuing (same device, same queue).
  inter-queue device-side enqueuing (same device, different queues).
  device-to-host enqueuing (device enqueue and host signal).
  multi-device device-side enqueuing (producer/consumer, doorbell signaling).
```

**Comment by benvanik-amd** [comment]
Another example idea: how to form the kernargs for HIP and OpenCL kernels. Most HSA docs don't mention this and AFAICT it's not documented anywhere but in code - I say that partly hoping I can be proven wrong and someone knows of some documentation :)

**Comment by atgutier** [comment]
> Another example idea: how to form the kernargs for HIP and OpenCL kernels. Most HSA docs don't mention this and AFAICT it's not documented anywhere but in code - I say that partly hoping I can be proven wrong and someone knows of some documentation :)

The HSA standard doesn't prescribe any specifics on how to implement support for the kernargs. It only prescribes what sort of visibility they should have (i.e., kernarg memory region properties).

**Comment by benvanik-amd** [comment]
Yep! It's less about what HSA expects and more about what OpenCL/HIP expect: since HSAIL is dead the only way to practically run anything on AMD GPUs with HSA is via those two and having the ABI documented/demonstrated would be useful.
